js/store.js

```diff
@@ -134,19 +256,74 @@ const store = {
     const firstParty = this.getWebsite(origin);
     const thirdParty = this.getWebsite(target);
 
+/*
+  if not {},
+
+  firstParty is expected to match this format:
+  {
+    thirdPartyHostnames: [
+      "www.thirdpartydomain.com",
+      ...
+    ]
+  }
+  thirdParty is expected to match this format:
+  {
+    firstPartyHostnames: [
+      "www.firstpartydomain.com",
+      ...
+    ],
+    isIgnored: true, (optional)
+    whiteListedFirstParty: "www.google.com" (optional)
+  }
+  */
+
+    let whiteListOverride = false;
+    let onWhiteList = false;
+
     if (!('thirdPartyHostnames' in firstParty)) {
```

Comment:
> you store state of isIgnored/isHidden/isVisible or whatever var name you choose

This is done already.

> you only return websites that are visible in getAll

This is now done:
```javascript
  getAll() {
    const output = {};
    for (const hostname in this._websites) {
      const website = this._websites[hostname];
      // only previously ignored websites will have an 'isVisible' key
      if (website['isVisible'] || !('isVisible' in website)) {
        output[hostname] = this.outputWebsite(hostname, website);
      }
    }
    return Promise.resolve(output);
  },
```



> you keep storing firstPartyHostnames in third party as this will never be seen until visible

Good point; this is done. Reverted this update back to its original logic:
```javascript
    if (!thirdParty['firstPartyHostnames'].includes(origin)) {
      thirdParty['firstPartyHostnames'].push(origin);
    }
```
> when this code decides it becomes visible, it should make sure all websites in firstPartyHostnames have a reference to the thirdParty in their thirdPartyHostnames arrays.

This is already done:
```javascript
// inside store.setThirdParty
} else if (!(this.onWhiteList(origin, target))) {
        // third party was previously whitelisted, and its
        // previously ignored link with the owning first party
        // should be added now that it links to a first party
        // for which it is not whitelisted
        // ... other stuff
        const whiteListedFirstParty = thirdParty['whiteListedFirstParty'];
        this._websites[whiteListedFirstParty]['thirdPartyHostnames']
          .push(target);
        // other stuff...
      }
```

This then gets written to storage when `setWebsite` is called.

> when a website becomes visible the update should fire, the update should not fire for this website until then

This is already done.

`newThirdParty` is only set to `true` if:
1. The third party is not whitelisted for the current request
2. The previously whitelisted third party has linked to a new first party for which it is not whitelisted

```javascript
// in store.setThirdParty
    if (newThirdParty) {
      this.updateChild(this.outputWebsite(target, data, origin));
    }
```

There is also a check in `store.setWebsite`:

```javascript
if (newSite && data['isVisible']) {
      this.updateChild(this.outputWebsite(hostname, websites[hostname]));
    }
```